flocker/control/test/test_persistence.py

```diff
@@ -249,3 +338,199 @@ def test_complex_keys(self):
+    links=st.sets(LINKS, max_size=10),
+    volume=st.none() | VOLUMES,
+    environment=st.dictionaries(keys=st.text(), values=st.text()),
+    memory_limit=NONE_OR_INT,
+    cpu_shares=NONE_OR_INT,
+    running=st.booleans()
+)
+
+
+def _build_node(applications):
+    # All the manifestations in `applications`.
+    app_manifestations = set(
+        app.volume.manifestation for app in applications if app.volume
+    )
+    # A set that contains all of those, plus an arbitrary set of
+    # manifestations.
+    dataset_ids = frozenset(
+        app.volume.manifestation.dataset_id
+        for app in applications if app.volume
+    )
+    manifestations = (
+        st.sets(MANIFESTATIONS.filter(
+            lambda m: m.dataset_id not in dataset_ids))
+        .map(pset)
+        .map(lambda ms: ms.union(app_manifestations))
+        .map(lambda ms: dict((m.dataset.dataset_id, m) for m in ms)))
+    return st.builds(
+        Node, uuid=UUIDS,
+        applications=st.just(applications),
+        manifestations=manifestations)
+
+
+NODES = st.sets(APPLICATIONS).map(
+    lambda apps: pset(dict((
+        app.volume.manifestation.dataset_id, app)
+        for app in apps if app.volume).values())
+).flatmap(_build_node)
+
+
+DEPLOYMENTS = st.builds(
+    Deployment, nodes=st.sets(NODES, min_size=0, max_size=3)
+)
+
+SUPPORTED_VERSIONS = st.integers(1, _CONFIG_VERSION)
+
+
+class ConfigurationMigrationTests(SynchronousTestCase):
+    """
+    Tests for ``ConfigurationMigration`` class that performs individual
+    configuration upgrades.
+
+    There should be
+    """
+    @given(DEPLOYMENTS)
+    def test_upgrade_configuration_v1_latest(self, deployment):
+        """
+        A range of generated configurations (deployments) can be
+        upgraded from v1 to the latest configuration version.
+
+        This test will need updating for each new configuration version
+        introduced to reflect the expected configuration format
+        (``expected_configuration``) after a successful upgrade.
+        """
+        source_json = wire_encode(deployment)
+        upgraded_json = migrate_configuration(
+            1, _CONFIG_VERSION, source_json, ConfigurationMigration)
+        upgraded_config = wire_decode(upgraded_json)
+        expected_configuration = Configuration(
+            version=_CONFIG_VERSION, deployment=deployment
+        )
+        self.assertEqual(upgraded_config, expected_configuration)
+
+    @given(st.tuples(SUPPORTED_VERSIONS, SUPPORTED_VERSIONS).map(
+        lambda x: tuple(sorted(x))))
+    def test_upgrade_configuration_versions(self, versions):
+        """
+        A range of versions can be upgraded and the configuration
+        blob after upgrade will matche that which is expected for the
+        particular version.
+
+        See flocker/control/test/configurations for individual
+        version JSON files and generation code.
+        """
+        configs_dir = FilePath(__file__).sibling('configurations')
```

Comment:
If you add a line:

```python
old_version, new_version = versions
```

And use those variables instead of `versions[0]` and `versions[1]`, the code will become magically more self-documenting.